### PR TITLE
`Exam mode`: Add quiz pool configuration

### PR DESCRIPTION
**Pull request by rriyaldhi**
#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/guidelines/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

Currently, if an instructor wants to have a high variance of quiz questions for an exam, the instructor needs to create a huge number of quiz exercises. Quiz Pool is created to improve the user experience for instructors to prepare such requirements by having groups of quiz questions that will be randomly selected for each student exam.

### Description

1. Add a new component QuizPoolMappingComponent to enable the instructor to map quiz questions to groups.
2. Add a new Quiz Pool Configuration page that uses the QuizPoolMappingComponent and QuizQuestionEditListComponent.
3. Add a **Quiz Pool** button to the Exam Detail Page
4. Add a new entity: QuizPool and QuizGroup.
5. Add the relationships between Exam, QuizQuestion, QuizGroup, and QuizPool.
6. Add Quiz Pool API to update and retrieve Quiz Pool Configuration.
7. Add quiz_pool and quiz_group Tables.
8. Add the relationships between exam, quiz_pool, quiz_question, and quiz_group Tables.

This is the class diagram that represents the changes. The classes and the lines with green color mean the classes and the relationships are added on this PR.
class

#### Exam Mode Testing

The feature is hidden by the latest commit, please use this build for testing: [https://bamboo.ase.in.tum.de/deploy/viewDeploymentVersion.action?versionId=624558853](https://bamboo.ase.in.tum.de/deploy/viewDeploymentVersion.action?versionId=624558853)

Prerequisites:
- 1 Instructor
- 1 Course
- 1 Exam

1. Log in to Artemis as an Instructor.
2. Navigate to the Exam.
3. Add **/quiz-pool** at the end of the url. (e.g. https://artemis-test1.artemis.in.tum.de/course-management/23/exams/483/quiz-pool)
4. Add or Import Quiz Questions.
11. Add Group(s).
12. Drag the Quiz Question Draggable Item to the Group.
13. Save the configuration and then refresh the page. Verify that the configuration is saved.
14. Drag the Quiz Question Draggable Item in a group to the unmap list.
15. Save the configuration and then refresh the page. Verify that the configuration is saved.
16. Delete a group, and verify that the Quiz Questions that were mapped to the group were moved to the unmap list.
17. Save the configuration and then refresh the page. Verify that the configuration is saved.
18. Back to the Exam checklist page and verify that there is a Quiz Exam entry on the exercise list table.
19. Click the **Quiz Pool** Button again and delete all questions and groups.
20. Back to the Exam checklist page and verify that the Quiz Exam entry on the exercise list table no longer exists. 

#### Performance Review
- [x] I confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Exam Mode Test
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [quiz-group.model.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4916-27/artifact/shared/Coverage-Report-Client-Tests/app/entities/quiz/quiz-group.model.ts.html) | 100% | ✅ |
| [quiz-pool.model.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4916-27/artifact/shared/Coverage-Report-Client-Tests/app/entities/quiz/quiz-pool.model.ts.html) | 100% | ✅ |
| [quiz-question.model.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4916-27/artifact/shared/Coverage-Report-Client-Tests/app/entities/quiz/quiz-question.model.ts.html) | 100% | ✅ |
| [exam-checklist-exercisegroup-table.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4916-27/artifact/shared/Coverage-Report-Client-Tests/app/exam/manage/exams/exam-checklist-component/exam-checklist-exercisegroup-table/exam-checklist-exercisegroup-table.component.ts.html) | 95.23% | ✅ |
| [exam-checklist.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4916-27/artifact/shared/Coverage-Report-Client-Tests/app/exam/manage/exams/exam-checklist-component/exam-checklist.component.ts.html) | 91.66% | ✅ |
| [quiz-pool-mapping-question-list.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4916-27/artifact/shared/Coverage-Report-Client-Tests/app/exercises/quiz/manage/quiz-pool-mapping-question-list.component.ts.html) | 100% | ✅ |
| [quiz-pool-mapping.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4916-27/artifact/shared/Coverage-Report-Client-Tests/app/exercises/quiz/manage/quiz-pool-mapping.component.ts.html) | 100% | ✅ |
| [quiz-pool.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4916-27/artifact/shared/Coverage-Report-Client-Tests/app/exercises/quiz/manage/quiz-pool.component.ts.html) | 92.92% | ✅ |
| [quiz-pool.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4916-27/artifact/shared/Coverage-Report-Client-Tests/app/exercises/quiz/manage/quiz-pool.service.ts.html) | 100% | ✅ |
| [navbar.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4916-27/artifact/shared/Coverage-Report-Client-Tests/app/shared/layouts/navbar/navbar.component.ts.html) | 90.87% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [QuizGroup.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4916-27/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.quiz/QuizGroup.html) | 47% | ✅ |
| [QuizPool.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4916-27/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.quiz/QuizPool.html) | 100% | ✅ |
| [QuizQuestion.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4916-27/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.quiz/QuizQuestion.html) | 97% | ✅ |
| [QuizPoolService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4916-27/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/QuizPoolService.html) | 97% | ✅ |
| [QuizPoolResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4916-27/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/QuizPoolResource.html) | 100% | ✅ |

### Screenshots

https://github.com/ls1intum/Artemis/assets/10885503/9b759c12-a72a-46f0-a09a-b1a5fab2d2e4

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/domain/exam/Exam.java
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/domain/exam/Exam.java:149]: I think this is a bit problematic, because it will be loaded `EAGER` by default, even if you specify it to be loaded `LAZY`, however it might not be necessary to do that in many cases.

File: src/main/java/de/tum/in/www1/artemis/domain/quiz/QuizQuestion.java
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/domain/quiz/QuizQuestion.java:74]: FetchType.LAZY does not work for ManyToOne. This will be loaded EAGER, which will lead to performance issues due to additional database queries, even if the value was never set. Can we avoid it?

File: src/main/java/de/tum/in/www1/artemis/domain/exam/Exam.java
Comment (rriyaldhi) [reply src/main/java/de/tum/in/www1/artemis/domain/exam/Exam.java]: This relation is removed.

File: src/main/java/de/tum/in/www1/artemis/domain/quiz/QuizQuestion.java
Comment (rriyaldhi) [reply src/main/java/de/tum/in/www1/artemis/domain/quiz/QuizQuestion.java:76]: The ManyToOne relation is removed and changed to QuizGroupId instead since we still need to map a QuizQuestion to a QuizGroup while saving the QuizPool.

**Review by frederik-kt — DISMISSED** [review]
Tested on TS2. UI looks really good. All dragging and dropping worked well.

**Review by laadvo — DISMISSED** [review]
Tested on ts2 during the testing session, quiz pool is working as expected :+1:

**Review by MaximilianAnzinger — DISMISSED** [review]
Tested on TS2 during the testing session.

**Review by tobias-lippert — APPROVED** [review]
Tested on ts2. I really liked the UI. Everything is working as expected.

**Review by pellumbbaboci — APPROVED** [review]
Tested during testing session on ts2, works as expected.

**Review by tobias-lippert — DISMISSED** [review]
Tested on ts2 in testing session . Everything worked

**Review by lennart-keller — APPROVED** [review]
Tested on ts2. Everything works as expected :)

**Comment by dakennguyen** [comment]
tested on ts1: for the exams that I created before I have deployed this PR, I can't see any options to choose
image

for new exam, everything works as expected

**Comment by rriyaldhi** [comment]
> tested on ts1: for the exams that I created before I have deployed this PR, I can't see any options to choose
> 
> image
> 
> 
> 
> for new exam, everything works as expected

Hi, could you check if the exam start date is in the future? Otherwise, the configuration is intentionally disabled.

File: src/test/javascript/spec/component/exercises/quiz/manage/quiz-pool-mapping-question-list.component.spec.ts
Comment (pellumbbaboci) [inline src/test/javascript/spec/component/exercises/quiz/manage/quiz-pool-mapping-question-list.component.spec.ts:59]: see GitHub Actions warnings

**Review by pellumbbaboci — DISMISSED** [review]
code looks good, i left a comment about some warnings from github actions

File: src/main/webapp/app/entities/quiz/quiz-group.model.ts
Comment (tobias-lippert) [inline src/main/webapp/app/entities/quiz/quiz-group.model.ts:4]: baseEntity already has an id, I think you can use the one defined there.

File: src/test/java/de/tum/in/www1/artemis/exam/QuizPoolIntegrationTest.java
Comment (tobias-lippert) [inline src/test/java/de/tum/in/www1/artemis/exam/QuizPoolIntegrationTest.java:104]: The teamscale warning might be addressed by creating a helper method that creates a quiz group with different parameters.

File: src/test/java/de/tum/in/www1/artemis/exam/QuizPoolIntegrationTest.java
Comment (MaximilianAnzinger) [inline src/test/java/de/tum/in/www1/artemis/exam/QuizPoolIntegrationTest.java:1]: You could break down the methods into smaller subtasks (creation of different objects for testing, or setup in general). This would improve readability and probably result in less LOC judging by the Teamscale findings.

File: src/test/javascript/spec/component/exercises/quiz/manage/quiz-pool-mapping.component.spec.ts
Comment (MaximilianAnzinger) [inline src/test/javascript/spec/component/exercises/quiz/manage/quiz-pool-mapping.component.spec.ts:68]:
```suggestion
        const addQuestion = jest.spyOn(component, 'addQuestion').mockImplementation();
```

File: src/test/javascript/spec/component/exercises/quiz/manage/quiz-pool-mapping.component.spec.ts
Comment (MaximilianAnzinger) [inline src/test/javascript/spec/component/exercises/quiz/manage/quiz-pool-mapping.component.spec.ts:118]: The name of this test seems odd to me

**Review by MaximilianAnzinger — CHANGES_REQUESTED** [review]
I added some comments. Other than what the others already mentioned during the testing session this looks good to me

File: src/main/webapp/app/entities/quiz/quiz-group.model.ts
Comment (MaximilianAnzinger) [inline src/main/webapp/app/entities/quiz/quiz-group.model.ts:4]:
```suggestion
```
`id` is defined in BaseEntity or do you explicitly require the `public` modifier? To my understanding, this should make no difference as `id` is public per default.

File: src/test/javascript/spec/component/exercises/quiz/manage/quiz-pool-mapping-question-list.component.spec.ts
Comment (rriyaldhi) [reply src/test/javascript/spec/component/exercises/quiz/manage/quiz-pool-mapping-question-list.component.spec.ts]: updated

File: src/main/java/de/tum/in/www1/artemis/repository/QuizGroupRepository.java
Comment (rriyaldhi) [reply src/main/java/de/tum/in/www1/artemis/repository/QuizGroupRepository.java:12]: added

File: src/main/java/de/tum/in/www1/artemis/web/rest/QuizPoolResource.java
Comment (rriyaldhi) [reply src/main/java/de/tum/in/www1/artemis/web/rest/QuizPoolResource.java]: updated

File: src/main/java/de/tum/in/www1/artemis/web/rest/QuizPoolResource.java
Comment (rriyaldhi) [reply src/main/java/de/tum/in/www1/artemis/web/rest/QuizPoolResource.java]: updated

File: src/main/java/de/tum/in/www1/artemis/web/rest/QuizPoolResource.java
Comment (rriyaldhi) [reply src/main/java/de/tum/in/www1/artemis/web/rest/QuizPoolResource.java]: updated

File: src/main/webapp/app/entities/quiz/quiz-group.model.ts
Comment (rriyaldhi) [reply src/main/webapp/app/entities/quiz/quiz-group.model.ts]: updated

File: src/main/webapp/app/entities/quiz/quiz-pool.model.ts
Comment (rriyaldhi) [reply src/main/webapp/app/entities/quiz/quiz-pool.model.ts]: updated

File: src/test/java/de/tum/in/www1/artemis/exam/QuizPoolIntegrationTest.java
Comment (rriyaldhi) [reply src/test/java/de/tum/in/www1/artemis/exam/QuizPoolIntegrationTest.java]: updated

File: src/test/javascript/spec/component/exercises/quiz/manage/quiz-pool-mapping-question-list.component.spec.ts
Comment (rriyaldhi) [reply src/test/javascript/spec/component/exercises/quiz/manage/quiz-pool-mapping-question-list.component.spec.ts]: updated

File: src/test/javascript/spec/component/exercises/quiz/manage/quiz-pool-mapping-question-list.component.spec.ts
Comment (rriyaldhi) [reply src/test/javascript/spec/component/exercises/quiz/manage/quiz-pool-mapping-question-list.component.spec.ts]: updated

File: src/test/javascript/spec/component/exercises/quiz/manage/quiz-pool-mapping.component.spec.ts
Comment (rriyaldhi) [reply src/test/javascript/spec/component/exercises/quiz/manage/quiz-pool-mapping.component.spec.ts]: updated

File: src/main/resources/config/liquibase/changelog/20230524102945_changelog.xml
Comment (rriyaldhi) [reply src/main/resources/config/liquibase/changelog/20230524102945_changelog.xml:12]: since this was approved and deployed to testing server, modifying this will break the PR. I added the default value to the java model instead.

File: src/test/java/de/tum/in/www1/artemis/exam/QuizPoolIntegrationTest.java
Comment (rriyaldhi) [reply src/test/java/de/tum/in/www1/artemis/exam/QuizPoolIntegrationTest.java:1]: updated

File: src/test/javascript/spec/component/exercises/quiz/manage/quiz-pool-mapping.component.spec.ts
Comment (rriyaldhi) [reply src/test/javascript/spec/component/exercises/quiz/manage/quiz-pool-mapping.component.spec.ts]: updated

File: src/main/java/de/tum/in/www1/artemis/service/QuizPoolService.java
Comment (rriyaldhi) [reply src/main/java/de/tum/in/www1/artemis/service/QuizPoolService.java:75]: yes, this is required for the case when a question had a group, but then the group is removed after that.

File: src/main/java/de/tum/in/www1/artemis/service/QuizPoolService.java
Comment (rriyaldhi) [reply src/main/java/de/tum/in/www1/artemis/service/QuizPoolService.java:25]: added

File: src/main/java/de/tum/in/www1/artemis/repository/QuizGroupRepository.java
Comment (rriyaldhi) [reply src/main/java/de/tum/in/www1/artemis/repository/QuizGroupRepository.java:12]: added

**Review by MaximilianAnzinger — DISMISSED** [review]
Thank you for addressing my suggestions.

**Review by tobias-lippert — APPROVED** [review]
Tested in testing session on ts1. Works as intended.

**Review by frederik-kt — DISMISSED** [review]
Tested on TS1. Works fine. 
Trying to add an empty name as a Quiz Pool returns the error that the name must be unique. This is minor but could be improved.

**Review by TimOrtel — DISMISSED** [review]
Tested on ts2. Creating and saving groups works as expected. Suggestion: The "add group" button should be disabled on illegal input.

**Review by MaximilianAnzinger — DISMISSED** [review]
Changes in code look good to me

**Review by tobias-lippert — APPROVED** [review]
Reapprove, code changes look good.

**Review by YeetTheFirst21 — DISMISSED** [review]
Tested on TS1,

all seems to be working as expected from the test steps.

**Review by tobias-lippert — DISMISSED** [review]
Quiz Pool button is hidden; code still looks good.

**Review by pellumbbaboci — APPROVED** [review]
the button is hidden now

**Review by MaximilianAnzinger — DISMISSED** [review]
Code LGTM.

Tested on TS1 during testing session. Looks good to me.

**Review by lennart-keller — DISMISSED** [review]
Retested on ts1, the quiz pool button is hidden now.

**Review by Santia-go — APPROVED** [review]
Deep tested on TS1. 
This partial implementation is working correctly.

**Review by YeetTheFirst21 — DISMISSED** [review]
Reapproved after merge conflict fix

**Review by MaximilianAnzinger — APPROVED** [review]
Reapproved after merge conflict

**Review by pellumbbaboci — DISMISSED** [review]
tested on ts2, works as expected 👍

**Review by Santia-go — APPROVED** [review]
Re-approved after solving merge conflict!

**Review by YeetTheFirst21 — DISMISSED** [review]
Reapproved after merge conflict fix

**Review by MaximilianAnzinger — DISMISSED** [review]
Reapproved after merge conflict

**Review by Santia-go — APPROVED** [review]
The merge conflict has been solved. 
Re-approved!

**Review by Santia-go — DISMISSED** [review]
Re-approved! 
The last changes do not affect the behaviour of the functionality previously tested.

File: src/test/java/de/tum/in/www1/artemis/exercise/quizexercise/QuizExerciseUtilService.java
Comment (MaximilianAnzinger) [inline src/test/java/de/tum/in/www1/artemis/exercise/quizexercise/QuizExerciseUtilService.java:877]: why are you returning the reference to the quiz. This is redundant.

**Review by MaximilianAnzinger — CHANGES_REQUESTED** [review]

File: src/test/java/de/tum/in/www1/artemis/exercise/quizexercise/QuizExerciseUtilService.java
Comment (rriyaldhi) [reply src/test/java/de/tum/in/www1/artemis/exercise/quizexercise/QuizExerciseUtilService.java]: Thanks for noticing!

**Review by MaximilianAnzinger — DISMISSED** [review]

**Review by YeetTheFirst21 — DISMISSED** [review]
Reapproved after resolve conflict

**Review by MaximilianAnzinger — APPROVED** [review]
Reapprove code after merge

**Review by YeetTheFirst21 — DISMISSED** [review]
Reapproved

**Review by lennart-keller — APPROVED** [review]
Retested the linked build on ts2. Worked fine.

**Review by Santia-go — APPROVED** [review]
Reapproved after the merge conflict has been solved!

**Review by YeetTheFirst21 — DISMISSED** [review]
Reapprove after Migration fixes

**Review by julian-christl — APPROVED** [review]
Tested on TS 1, works as expected.

I recommend adding a third section "assigned to everyone" instead of defaulting back to that behaviour. It's very unintuitive. But that can be done in a follow up. We discussed pros and cons in the testing session

**Review by florian-glombik — DISMISSED** [review]
Tested during testing session on ts1 - the UI works fine, good job!

Here are some minor issues / improvement suggestions that can be addressed in a follow up PR:
- the separation between _Edit Quiz Pool_ and _Questions_ could be clearer, e.g. indicated by a margin at the left
image
- These Info Icons usually are used to indicate some information via hover in Artemis, so I would prefer a consistent usage -> make 1 a tooltip, remove icon 2 and just write the text?
image
- The add group button could be wrapped later
image

File: src/main/webapp/app/exam/manage/exams/exam-checklist-component/exam-checklist.component.ts
Comment (MaximilianAnzinger) [inline src/main/webapp/app/exam/manage/exams/exam-checklist-component/exam-checklist.component.ts:65]: How is `404` not an error and doesn't cause an alert? What could cause an `404`-error that should be handled by ignoring it?

**Review by MaximilianAnzinger — COMMENTED** [review]
Code LGTM, I just have one question.

File: src/main/webapp/app/exam/manage/exams/exam-checklist-component/exam-checklist.component.ts
Comment (rriyaldhi) [reply src/main/webapp/app/exam/manage/exams/exam-checklist-component/exam-checklist.component.ts:65]: In this case, it is not an error from the client perspective because it is possible that an exam (e.g. old exam) doesn’t have a quiz pool entity. This error is not ignored but it is handled by initializing a new quiz pool entity.

**Review by MaximilianAnzinger — DISMISSED** [review]

**Review by Kroko-fant — APPROVED** [review]
tested during the testing session on ts1 worked fine so far apart from some ui improvements as mentioned by florian/julian

**Review by Santia-go — APPROVED** [review]
Tested on TS3 
Working as expected after the merge conflict was solved. 
Agree with the other testers that some UI improvements would be very suitable in follow up PRs. 
Approved!

**Review by pal03377 — DISMISSED** [review]
Tested on TS2 (testing session), everything worked as expected. Small issue for a follow-up: When a group name starts with `<`, the error message breaks:

screenshot-2023-09-28_001988

**Review by MaximilianAnzinger — APPROVED** [review]
Reapprove code changes
Tested on TS2 besides the issue mentioned by paul everything looks good

**Review by lennart-keller — APPROVED** [review]
Tested on TS2 during testing session. Worked well (except for the one little bug).

**Review by YeetTheFirst21 — DISMISSED** [review]
Reapproved after migration conflicts

**Review by Santia-go — APPROVED** [review]
Tested on TS1 
Working as expected!

**Pull request merged**
